plugins/plugin-ssh-machine/src/main/java/org/eclipse/che/plugin/machine/ssh/SshMachineInstance.java

```diff
@@ -0,0 +1,199 @@
+
+import com.google.inject.assistedinject.Assisted;
+
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.model.machine.Command;
+import org.eclipse.che.api.core.model.machine.Machine;
+import org.eclipse.che.api.core.model.machine.ServerConf;
+import org.eclipse.che.api.core.util.LineConsumer;
+import org.eclipse.che.api.machine.server.exception.MachineException;
+import org.eclipse.che.api.machine.server.model.impl.MachineRuntimeInfoImpl;
+import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
+import org.eclipse.che.api.machine.server.spi.Instance;
+import org.eclipse.che.api.machine.server.spi.InstanceKey;
+import org.eclipse.che.api.machine.server.spi.InstanceNode;
+import org.eclipse.che.api.machine.server.spi.InstanceProcess;
+import org.eclipse.che.api.machine.server.spi.impl.AbstractInstance;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.ws.rs.core.UriBuilder;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Implementation of {@link Instance} that uses represents ssh machine.
+ *
+ * @author Alexander Garagatyi
+ *
+ * @see SshMachineInstanceProvider
+ */
+// todo try to avoid map of processes
+public class SshMachineInstance extends AbstractInstance {
+    private static final AtomicInteger pidSequence = new AtomicInteger(1);
+
+    private final SshClient                                   sshClient;
+    private final LineConsumer                                outputConsumer;
+    private final SshMachineFactory                           machineFactory;
+    private final Set<ServerConf>                             machinesServers;
+    private final ConcurrentHashMap<Integer, InstanceProcess> machineProcesses;
+
+    private MachineRuntimeInfoImpl machineRuntime;
+
+    @Inject
+    public SshMachineInstance(@Assisted Machine machine,
+                              @Assisted SshClient sshClient,
+                              @Assisted LineConsumer outputConsumer,
+                              SshMachineFactory machineFactory,
+                              @Named("machine.ssh.machine_servers") Set<ServerConf> machinesServers) {
+        super(machine);
+        this.sshClient = sshClient;
+        this.outputConsumer = outputConsumer;
+        this.machineFactory = machineFactory;
+        this.machinesServers = machinesServers;
+        this.machineProcesses = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public LineConsumer getLogger() {
+        return outputConsumer;
+    }
+
+    @Override
+    public MachineRuntimeInfoImpl getRuntime() {
+        // lazy initialization
+        if (machineRuntime == null) {
```

Comment:
If `getRuntime` method may be used from several threads in the same time then you have to use double-checked-volatile idiom here